libmuscle/python/libmuscle/mmsf_validator.py

```diff
@@ -52,36 +52,37 @@ def __init__(self, port_manager: PortManager) -> None:
 
         # Allowed operator transitions, the following are unconditionally allowed:
         self._allowed_transitions = {
-                (Operator.NONE, Operator.NONE),
-                (Operator.NONE, Operator.F_INIT),
-                (Operator.F_INIT, Operator.O_I),
-                (Operator.F_INIT, Operator.O_F),
-                (Operator.O_I, Operator.S),
-                (Operator.S, Operator.O_I),
-                (Operator.S, Operator.O_F),
-                (Operator.O_F, Operator.NONE),
-                }
+                Operator.NONE: [Operator.NONE, Operator.F_INIT],
+                Operator.F_INIT: [Operator.O_I, Operator.O_F],
+                Operator.O_I: [Operator.S],
+                Operator.S: [Operator.O_I, Operator.O_F],
+                Operator.O_F: [Operator.NONE]}
         # If there are operators without connected ports, we can skip over those
         for operator in [Operator.F_INIT, Operator.O_I, Operator.S, Operator.O_F]:
             if not self._connected_ports.get(operator, []):
                 # Find all transitions A -> operator -> B and allow transition A -> B:
```

Comment:
> Maybe this comment could be extended to point out that it's transitive to avoid future confusion?

Sure, will do

> Also, in that case, will we end up with None -> O_I in _allowed_transitions, despite O_I not having connected ports? That's probably not a problem, since that transition cannot occur anyway?

Correct, and it is deliberately so. Users can still send on an O_I port when it is not connected (the Communicator will just discard the message). For the MMSFValidator this transition still happens and is valid :)